Perplex lookup volume fractions

bobmyhill opened the pull request:
New functionality: The PerpleX lookup now allows the user to add columns to the lookup which correspond to the volume fractions of phases in the assemblages. These can then be read by material models. 

This PR also provides an example of the use of the new functionality via a modification to the Steinberger material model. The material model calculates the phase volume proportions and passes them to a NamedAdditionalMaterialOutputs object.

Finally, a new test is provided which checks the new functionality.

This PR builds on PR #3707 (now merged).

gassmoeller commented on include/aspect/material_model/interface.h:
```diff
@@ -903,6 +903,31 @@ namespace aspect
+     * MaterialModel::Interface::evaluate() function.
+     */
+    template <int dim>
+    class PhasePropertyOutputs : public NamedAdditionalMaterialOutputs<dim>
```
Actually I think this is general enough to be an addition to the base class `NamedAdditionalMaterialOutputs` and should be an independent pull request. The only modification to the base class is:
- Add a constructor to the base class that takes `property_names` and `n_points` and is implemented exactly like you do here
- Add a ` std::vector<std::vector<double> > property_values;` member to the base class
- Add a `get_nth_output` implementation to the base class that does exactly what you do here, except it also contains an `Assert` that the `property_values.size` is actually set to something > 0 (to make sure it was properly initialized).

Then you can simply use an object of type `NamedAdditionalMaterialOutputs` instead of `PhasePropertyOutputs`. And in the future the other derived classes could use the new functionality as well (they can keep their names if they want, but use the functionality you implemented here). I think this simplifies the implementation.

bobmyhill replied:
Resolved by #3772

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -383,6 +409,56 @@ namespace aspect
+            }
+
+        }
+      else if (material_lookup.size() == compositional_fields.size() + 1)
```
Have you tested that this works with more than one compositional field, and different phase names for the different compositions, with just some overlap?

bobmyhill replied:
Now I have :)

bobmyhill commented on data/material-model/steinberger/pyr_with_volume_fractions.txt:
```diff
@@ -0,0 +1,1613 @@
+   56.410256410256409
+          40
+          30
+P(bar)         T(K)           rho,kg/m3      alpha,1/K      cp,J/K/kg      vp,km/s        vs,km/s        h,J/kg        	vol_Pl1 vol_Pl2 vol_O vol_Cpx1 vol_Cpx2 vol_ky vol_q vol_Opx vol_Gt1 vol_Gt2 vol_Sp vol_C2overc vol_Ring vol_Wad vol_st vol_Wus1 vol_Wus2 vol_Aki1 vol_Aki2 vol_capv vol_CF vol_Pv
```
Actually, this isn't the raw PerpleX file; I had to run the output through a (bash) script to get everything in the correct format. I guess I should add the script to contrib? Or is there a better place for it?

The names are those that come out of PerpleX; capitalized phases signify that the phase is a solution of different endmembers, lower case are pure endmembers. I can change the columns by hand if you want?

bobmyhill commented on source/material_model/interface.cc:
```diff
@@ -914,6 +914,32 @@ namespace aspect
+      // we have to extract the nth phase property output
+      // but the vector in the material model outputs is sorted so that the
+      // number of evaluation points (and not the number of phase properties)
+      // is the outer vector
```
At the moment, the Steinberger model loops over each evaluation point, so it was easier to pass in.temperature[i], in.pressure[i], in.composition[i] ... to the new phase_volume_fraction function. 

But I could equally pass the function a reference to `in` and do both the inner and outer loops within the function. This will mean a few more lines of code in Steinberger, but interface.cc will be neater (and a bit faster).

bobmyhill replied:
Now implemented the other way round :)

bobmyhill commented on include/aspect/material_model/steinberger.h:
```diff
@@ -305,6 +310,12 @@ namespace aspect
+        * List of strings containing the names of the unique phases in all the material lookups
+        */
+        std::vector<std::string> unique_phase_names;
+        std::vector<std::vector<int>> unique_phase_indices;
```
Now an unsigned int.
These have to be in the Steinberger model because unique_phase_indices is potentially a function of all of the compositional fields read into Steinberger.

bobmyhill commented on include/aspect/material_model/interface.h:
```diff
@@ -903,6 +903,31 @@ namespace aspect
+         * that are passed to the instance of MaterialModel::Interface::evaluate()
+         * that fills the current object.
+         * phase_property_values[q][c] is the phase property at the evaluation point q
+         * for the compositional field with the index c.
```
Now outdated, after removing PhasePropertyOutputs :)

bobmyhill commented on source/material_model/steinberger.cc:
```diff
@@ -132,16 +132,42 @@ namespace aspect
     void
     Steinberger<dim>::initialize()
     {
+      std::set<std::string> set_phase_names;
```
This is used to make a set of all of the phase names from all of the lookups. Some phase names will be repeated in all/some of the lookups, others will already be unique.

Making a unique_phase_names set and then making an array which maps the correct index in each lookup to the index in unique_phase_names saves on looking things up again and again.

I'll add some comments here making all of this more explicit.

bobmyhill commented on source/material_model/steinberger.cc:
```diff
@@ -132,16 +132,42 @@ namespace aspect
         = std_cxx14::make_unique<internal::RadialViscosityLookup>(data_directory+radial_viscosity_file_name,
                                                                   this->get_mpi_communicator());
       avg_temp.resize(n_lateral_slices);
+
```
Yes :)

bobmyhill commented on include/aspect/material_model/utilities.h:
```diff
@@ -156,6 +167,10 @@ namespace aspect
 
+            std::vector<std::string> phase_names;
+            std::map<std::string, int> phase_name_index;
+            std::vector<dealii::Table<2,double>> phase_volume_fraction_values;
```
Good idea.

bobmyhill commented on include/aspect/material_model/utilities.h:
```diff
@@ -156,6 +167,10 @@ namespace aspect
             dealii::Table<2,double> enthalpy_values;
 
+            std::vector<std::string> phase_names;
+            std::map<std::string, int> phase_name_index;
```
Ooops, missed this one. I agree that we only need either a pair of vectors or a map, but having both is faster, I think. 
- Converting all the keys into a vector every time `phase_volume_column_names()` is called is a bit pointless.
- Running a `find_if` on `phase_column_names` to get the index of a string, to then use to get an int from another vector is also complicated when compared with using a map.

Am I missing a third way?

bobmyhill replied:
I've changed the name of the map to phase_name_index_map anyway, whatever we end up choosing.

bobmyhill replied:
I've removed the map and changed the logic of the phase column lookup in steinberger.cc.

bobmyhill commented on include/aspect/material_model/utilities.h:
```diff
@@ -118,6 +118,17 @@ namespace aspect
                     const double pressure) const;
 
+            std::vector<std::string>
+            phase_volume_phase_names() const;
```
Done!

bobmyhill commented on include/aspect/material_model/steinberger.h:
```diff
@@ -183,6 +183,11 @@ namespace aspect
+        virtual std::vector<double> phase_volume_fractions (const double      temperature,
+                                                            const double      pressure,
+                                                            const std::vector<double> &compositional_fields,
+                                                            const Point<dim> &position) const;
```
The updated NamedAdditionalMaterialOutputs now requires that we pass a reference to in, so this has now changed.

bobmyhill commented on source/material_model/steinberger.cc:
```diff
@@ -383,6 +409,56 @@ namespace aspect
+                }
+            }
+        }
+      else
```
Good catch.

jdannberg commented on source/material_model/utilities.cc:
```diff
@@ -538,29 +552,76 @@ namespace aspect
+                                                 "pressure and temperature, so you may see several columns with the same name. "
+                                                 "Either combine columns with the same name, or change the names."));
+                        }
+                      else
```
You  don't have to put this into an else statement. If the assert triggers, we never get here anyway.

bobmyhill replied:
:D

jdannberg commented on source/material_model/utilities.cc:
```diff
@@ -538,29 +552,76 @@ namespace aspect
+                        {
+                          // Populate phase_column_names with the column name
+                          // and phase_column_indices with the column index in the current lookup file
+                          // This makes populating the phase_volume_fractions object easier.
```
Can you explain this a bit more? (Why? Easier than what?)

bobmyhill replied:
I've removed this line; it used to make sense when I had the map, but not any more!

jdannberg commented on source/material_model/utilities.cc:
```diff
@@ -618,13 +683,34 @@ namespace aspect
+                      phase_volume_fractions[n][i%n_temperature][i/n_temperature]=row_values[phase_column_indices[n]];
+                    }
+                }
+              else // first_natural_variable == "P(bar)"
```
Wait, was this here before? What does it do? Is that change related to the other changes in this pull request?

bobmyhill replied:
No, it wasn't here before - this was the headache I found on Wednesday night.

The ordering of the first two columns in the PerpleX table files dictates whether the inner loop is over temperature or pressure. The first column is always the inner loop.

I've now added some documentation to describe this. The existing tests cover the possible range of possibilities.

jdannberg replied:
I am just saying: this is independent of the phase stuff, right? Before, we just didn't allow having pressure as the inner loop? We could also just make sure temperature is always the first column. Is there a reason why one would want to have pressure first?

jdannberg replied:
But I guess since it's already part of this pull request, and you've made a test for it, if there's a reason why someone would want to have pressure first, let's just leave this as it is.

bobmyhill replied:
Thanks; yes, I should have split this into a separate PR, but now that it's in there...

The reason why I think this is needed is that I genuinely don't understand how PerpleX chooses how to output the table. The ordering of P-first or T-first seems to have changed a few months ago, or maybe there's a flag that I don't know about.

Either way, if I don't know about how PerpleX makes these choices, I'm sure there are other users who would be similarly PerpleXed if their completely-reasonable-looking input file produced garbage. It also makes the code somewhat more future-proof.

jdannberg commented on tests/steinberger_phase_fractions_background.prm:
```diff
@@ -0,0 +1,131 @@
+# computed. The property table also contains volume fractions of the
+# stable phases in the rock, which can be output using
+# the "named additional outputs" in the list of named outputs for
+# the Visualization postprocessor.
```
Can you update the description to make it clear how this test is different from the one above?

bobmyhill replied:
Done.

jdannberg commented on include/aspect/material_model/utilities.h:
```diff
@@ -118,6 +118,23 @@ namespace aspect
+            /**
+             * Returns the volume fraction of the phase_idth phase
+             * at a given temperature and pressure, where the phase_column_name
+             * corresponds to that given in phase_volume_column_names()[phase_id].
```
Wait, but this doesn't really have anything to do with what phase_volume_column_names() does, phase_volume_column_names() could be changed, but this function would still return the volume fraction of the phase with the column index phase_id at a given temperature and pressure. I think the thing you're saying here really belongs into the documentation of std::vector<std::string> phase_column_names and std::vector<dealii::Table<2,double>> phase_volume_fractions (namely, that these belong together and are in the same order).

bobmyhill replied:
Good point. I've removed this sentence and added documentation to phase_column_names and phase_volume_fractions.

bobmyhill commented on contrib/python/create_aspect_readable_perplex_table_with_phase_volumes.py:
```diff
@@ -0,0 +1,180 @@
+--n_pressures N_PRESSURES
+--n_temperatures N_TEMPERATURES
+[--pressure_range LOW_PRESSURE_LIMIT HIGH_PRESSURE_LIMIT]
+[--temperature_range LOW_TEMPERATURE_LIMIT HIGH_TEMPERATURE_LIMIT]
```
This is what comes out of the python argument parser <shrugs>

bobmyhill commented on source/material_model/steinberger.cc:
```diff
@@ -383,6 +426,74 @@ namespace aspect
+            {
+              for (unsigned int j = 0; j < in.n_evaluation_points(); j++)
+                {
+                  if (unique_phase_volume_column_indices[0][i] != -1)
```
I had the realisation this morning that the vector of vector<double> doesn't need to look like a regular table, and that therefore I could reverse the logic of the lookup, removing all of the -1s and speeding up the lookups...

bobmyhill replied:
So I've now done that. Sorry for making you look through another iteration of this, but I think this is closer to what you would like and is cleaner.

bobmyhill commented on data/material-model/steinberger/morb_G13_with_volume_fractions_lo_res.dat:
```diff
@@ -0,0 +1,2529 @@
+|6.6.6
```
Yes, and also for pyrolite. I've put the references and compositions into the prm files.

bobmyhill commented on include/aspect/material_model/steinberger.h:
```diff
@@ -305,6 +307,12 @@ namespace aspect
+        * List of strings containing the names of the unique phases in all the material lookups
+        */
+        std::vector<std::string> unique_phase_volume_column_names;
+        std::vector<std::vector<int>> unique_phase_volume_column_indices;
```
Done, and done.

bobmyhill commented on include/aspect/material_model/steinberger.h:
```diff
@@ -183,6 +183,8 @@ namespace aspect
                                    const std::vector<double> &compositional_fields,
                                    const Point<dim> &position) const;
 
+        virtual std::vector<std::vector<double>> phase_volume_fractions (const MaterialModel::MaterialModelInputs<dim> &in) const;
```
Done.

bobmyhill commented:
Ok, I'm not sure why the new tests are failing. I'll investigate after morning rounds.

jdannberg commented:
Okay, let me know when you want me to look at this again!

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -489,14 +600,20 @@ namespace aspect
+      // fill phase volume outputs if they exist
+      if (NamedAdditionalMaterialOutputs<dim> *phase_volume_fractions_out = out.template get_additional_output<NamedAdditionalMaterialOutputs<dim> >())
+        {
+          phase_volume_fractions_out->output_values = phase_volume_fractions(in);
```
What about this? Didn't work, or you just missed my comment?

bobmyhill replied:
Missed this one, but also my attempts over the past hour and a bit didn't work. Can we raise this as an issue for later?

bobmyhill replied:
I should elaborate: I changed the function, gave it std::vector<std::vector<double>> &volume_fractions as a new argument, and passed phase_volume_fractions_out->output_values to the function.

The vector of vectors filled fine within the function, and I could retrieve the values within the if {} clause, but then maybe the values got lost and ASPECT seg-faulted during postprocessing.

bobmyhill replied:
I missed your comment originally, and then had problems last night doing what you suggested. It works this morning though :)

bobmyhill commented on source/material_model/steinberger.cc:
```diff
@@ -132,9 +132,52 @@ namespace aspect
+              unique_phase_volume_column_indices[i][i_unique] = j;
+            }
+        }
+
```
Github had this comment, and I didn't see it. Thank goodness I had the same idea :)

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -383,6 +426,74 @@ namespace aspect
+            }
+        }
+      else if (material_lookup.size() == in.composition[0].size() + 1) // background field
+        {
```
I still think it would be easier to read without the brackets :grin:

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -383,6 +427,77 @@ namespace aspect
+              for (unsigned int k = 0; k < in.n_evaluation_points(); k++)
+                {
+                  background_volume_fractions[unique_phase_indices[0][j]][k] += material_lookup[0]->phase_volume_fraction(j,in.temperature[k],in.pressure[k]);
+                  volume_fractions[unique_phase_indices[0][j]][k] += background_volume_fractions[unique_phase_indices[0][j]][k];
```
I think you can avoid looping over j and k twice, if you loop from i=0 below and then just make this the spacial case for i=0.

bobmyhill replied:
Ok!

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -132,9 +132,53 @@ namespace aspect
+
+      // Initialise the column indices
+      unique_phase_indices.resize(material_file_names.size(),
+                                  std::vector<unsigned int>());
```
I would put that on one line. Also: you can resize this before you start your first loop in line 144, and then you only need one loop over all i.

bobmyhill replied:
I don't think this is true. The second loop over i requires `unique_phase_names`, which is only complete at the very end of the first loop over i.

bobmyhill replied:
The call to resize is now on one line.

bobmyhill replied:
Oh, wait, now I see what you mean. When this involved a set, I needed the two loops, but now I don't. Very good catch!

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -132,9 +132,53 @@ namespace aspect
+      for (unsigned int i = 0; i < material_file_names.size(); i++)
+        {
+          std::vector<std::string> phase_volume_column_names = material_lookup[i]->phase_volume_column_names();
+          for (unsigned int j = 0; j < phase_volume_column_names.size(); j++)
```
Same here: you now loop over all j twice, but you can combine this with the loop over the strings above. (At least I am thinking this will make the code shorter, if that just makes it more complicated, it's not worth changing).

bobmyhill replied:
Same problem as above. The two loops need to be separated, because the second loop requires a completed `unique_phase_names` object from the first.

bobmyhill replied:
Not a problem!

jdannberg commented on source/material_model/steinberger.cc:
```diff
@@ -132,9 +132,53 @@ namespace aspect
+            {
+              std::vector<std::string>::iterator it = std::find(unique_phase_names.begin(),
+                                                                unique_phase_names.end(),
+                                                                phase_volume_column_names[j]);
```
The finding you also only need to do once

bobmyhill replied:
Yup, now changed. Thanks!

bobmyhill commented on source/material_model/steinberger.cc:
```diff
@@ -383,6 +427,77 @@ namespace aspect
+    {
+      // In the following function,
+      // the index i corresponds to the ith compositional field
+      // the index j corresponds to the jth unique phase
```
Right. Good catch.

bobmyhill commented:
So it turns out that trying to fix `things` late at night is not a productive use of time. Who knew?
I think I've addressed all your comments. The code looks a lot neater now, thank you!

jdannberg commented on include/aspect/material_model/steinberger.h:
```diff
@@ -183,6 +183,14 @@ namespace aspect
+        /**
+        * Returns the volume fractions of all the phases over all of the current
+        * evaluation points. The outer loop is over the phases, and the inner
+        * loop is over all the evaluation points.
```
You still need to update the documentation. (the function doesn't return the volume fractions any more). 
I would also phrase the second part in a different way: You want to say what this new input object is and how it is filled. Something like: 
This function uses the MaterialModelInputs &in to fill the output_values of the phase_volume_fractions_out output object with [...]. This output_values object is a vector of vectors; the outer vector is expected to have a size that equals the number of unique phases, the inner vector is expected to have a size that equals the number of evaluation points.

jdannberg commented:
Great, thank you!
